File: challenger-backend/src/main/java/com/kameo/challenger/odb/ChallengeODB.java
package com.kameo.challenger.odb;


import com.google.common.collect.Lists;
import com.kameo.challenger.odb.api.IIdentity;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.ToString;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.util.Date;
import java.util.List;

@Entity
@NoArgsConstructor
@ToString(of = "id")
public @Data class ChallengeODB implements IIdentity {
    @Id
    @GeneratedValue(strategy= GenerationType.AUTO)
    long id;

    String label;

    @NotNull
    @ManyToOne
    UserODB createdBy;


    @Override
    public long getId() {
        return id;
    }

    @NotNull
    @Enumerated
    ChallengeStatus challengeStatus;


    @OneToMany(mappedBy = "challenge")
    List<ChallengeParticipantODB> participants= Lists.newArrayList();


    public ChallengeODB(long id) {
        this.id = id;
    }

    public int hashCode() {
        return Long.hashCode(id);
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof ChallengeODB && ((ChallengeODB) obj).getId()==this.getId();
    }

    @Override
    public boolean isNew() {
        return getId()<0;
    }
}

File: challenger-backend/src/main/java/com/kameo/challenger/domain/challenges/ChallengeODB.kt
package com.kameo.challenger.domain.challenges

import com.google.common.collect.Lists
import com.kameo.challenger.domain.challenges.ChallengeParticipantODB
import com.kameo.challenger.domain.challenges.ChallengeStatus
import com.kameo.challenger.odb.UserODB
import com.kameo.challenger.odb.api.IIdentity
import javax.persistence.*
import javax.validation.constraints.NotNull

@Entity
data class ChallengeODB(@Id
                        @GeneratedValue(strategy = javax.persistence.GenerationType.AUTO)
                        override var id: Long = 0) : IIdentity {

    var label: String? = null

    @ManyToOne
    lateinit var createdBy: UserODB;

    @Enumerated
    lateinit var challengeStatus: ChallengeStatus;

    @OneToMany(mappedBy = "challenge")
    var participants: List<ChallengeParticipantODB> = Lists.newArrayList<ChallengeParticipantODB>()
}